pymongo/client_session.py

```diff
@@ -368,6 +377,97 @@ def _inherit_option(self, name, val):
             return val
         return getattr(self.client, name)
 
+    def with_transaction(self, callback, read_concern=None, write_concern=None,
```

Comment:
The spec allows some wiggle room in the API we expose for `callback`. For now, I've implemented the `callback` as a function that accepts a single argument, the session.

Should we add support for passing arbitrary options into the callback?:
```python
def callback(session, custom_arg, custom_kwarg=None):
    orders = session.client.db.orders
    inventory = session.client.db.inventory
    orders.insert_one({"sku": "abc123", "qty": 100}, session=session)
    inventory.update_one({"sku": "abc123", "qty": {"$gte": 100}},
                         {"$inc": {"qty": -100}}, session=session)


with client.start_session() as session:
    session.with_transaction(callback, 'custom_arg', custom_kwarg=10)
```

I think that would be handy but I'm not quite sure how to get such an API and also support the necessary transaction options. Also, if there is a way to do this, then we will have to worry about breaking compatibility anytime we introduce a new argument to `with_transaction`, for example when we add a new TransactionOption.

One way to have the "best" of both worlds would be:
```python
def with_transaction(self, callback, args=None, kwargs=None,
                     read_concern=None, write_concern=None,
                     read_preference=None):
     # ...
     args = (,) if args is None else args
     kwargs = {} is kwargs is None else kwargs
     ret = callback(self, *args, **kwargs)
     # ...

# Example:
with client.start_session() as session:
    session.with_transaction(callback, args=('custom_arg',),
                             kwargs={'custom_kwarg': 10})
```
But this doesn't seem very pythonic to me. 

After writing this out I think I'm somewhat convinced that the current API is good enough. If the user wants a callback to accept various arguments they can wrap the callback in a lambda like this:
```python
with client.start_session() as session:
    session.with_transaction(
        lambda s: callback(s, 'custom_arg', custom_kwarg=10))
```
In this case it's a bit cumbersome to pass the session (`s`) because it's already in scope but I think that's fine.

Any thoughts?

Reply:
Your idea of accepting args and kwargs isn't totally unprecedented in Python -  https://docs.python.org/2/library/threading.html#threading.Thread

Reply:
I like the args kwargs idea. Also, we'll hopefully get some beta feedback on this and can change our minds before GA.

Reply:
Same goes for `multiprocessing.Process`: `class multiprocessing.Process(group=None, target=None, name=None, args=(), kwargs={})`. So it definitely seems like this is not an antipattern for Python.

Reply:
Interesting, I didn't realize this pattern was already in use in the standard library. We can easily add support for `args=(), kwargs={}` in a non-breaking way but we can't remove support once we release it. So, I propose we not add support for it now and wait to see if users think the current api is too restrictive.

Reply:
Good point about not being able to introduce the current API without a backward breaking change. But it also got me thinking that the standard library API is a superset of what is currently implemented so why not just implement it? Users are free to use `functools.partial` (or something similar) if they so desire. Just a thought!

Reply:
I'll open a new ticket to add `args`, `kwargs`. I'm starting to like the idea but I still want to hold off until we get more feedback from real users.